addons/test_mail/tests/test_attachment.py

```diff
@@ -0,0 +1,72 @@
+from odoo.addons.test_mail.tests.common import TestMailCommon
+from odoo.exceptions import AccessError
+
+class TestAttachment(TestMailCommon):
+
+    def test_attachment_forbid_unlink(self):
+        """Check that removing message attachments is prevented on other user's messages."""
+        def write(attachment, w_user, w_sudo=False):
+            attachment.with_user(w_user).sudo(w_sudo).write({'name': f'{w_user.name} {attachment.name[-1]}'})
+
+        def unlink(attachment, w_user, w_sudo=False):
+            attachment.exists().with_user(w_user).sudo(w_sudo).unlink()
+
+        test_record = self.env['mail.test.simple'].with_context(self._test_context).create({
+            'name': 'Test',
+            'email_from': 'ignasse@example.com',
+        })
+
+        user_admin, user_employee = self.env['res.users'].search([('login', 'in', ['admin', 'employee'])], order='login asc')
+        admin_attachments = self.env['ir.attachment'].with_user(user_admin).create([{
+            'type': 'binary',
+            'name': f'admin_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(2)])
+        employee_attachments = self.env['ir.attachment'].with_user(user_employee).create([{
+            'type': 'binary',
+            'name': f'employee_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(3)])
+
+        user_employee_clone = user_employee.copy(default={'login': 'employee_clone', 'email': 'employee_clone@local.lan'})
+        employee_clone_attachments = self.env['ir.attachment'].with_user(user_employee_clone).create({
+            'type': 'binary',
+            'name': 'employee_clone_attachment',
+            'datas': "ABC=",
+        })
+
+        test_record.with_user(user_admin).message_post(body="Hi", attachment_ids=admin_attachments.ids)
+        test_record.with_user(user_employee).message_post(body="Hello", attachment_ids=employee_attachments.ids)
+        test_record.with_user(user_employee_clone).sudo().message_post(body="Hello again", attachment_ids=employee_attachments[2].ids)
+        create_by_employee_send_by_clone = employee_attachments[2]
+        test_record.with_user(user_employee_clone).sudo().message_post(body="Hello again with own attachment", attachment_ids=employee_clone_attachments.ids)
+
+        # forbidden
+        forbidden_list = [
+            (user_employee, admin_attachments[0], False),
+            (user_employee, employee_clone_attachments[0], False)
+        ]
+        for method in (write, unlink):
```

Comment:
Personnally I would instead do two subTest and explicitely call write and unlink. Not mandatory to change, just that I feel like this hides a bit what the test is doing.

Reply:
Sure as I do it further down anyway. Idea was just to reduce the noise from the test since here it's obvious we're testing the same behavior twice, just with different methods.

Will still apply the change as maybe it's not as easy to understand what's going on in the end.